### DISCUSSION TITLE
[Kavita] Alternative account provider (OIDC/SSO)

### DISCUSSION THREAD

**Discussion by DieselTech** [Kareadita/Kavita#2533] [Ideas]
### Idea Description

Posted by BrainWart
· Nov 2021
· 122 Combined upvotes between:

https://feats.kavitareader.com/posts/96/add-openid-connect-support-or-allow-user-login-by-header
https://feats.kavitareader.com/posts/23/alternative-account-provider

Description

I would like to have users sign in using OAuth or some other authentication provider to help with centralized account management.

### Duration of Using Kavita

42

### Idea Category

Feature Enhancement

### Agreement

- [X] I agree that this is solely for submitting ideas, and I will search for existing ideas before posting.

**Comment by earth4s** [comment]
It would be nice to give my users an option to login with their Google accounts, this might take some of the burden of account and password management off the admins plate.

**Comment by DDriggs00** [comment]
I definitely agree that this is an important feature, especially with how useful SSO systems like Authentik, Authelia, etc. are.

**Comment by nyhil116** [comment]
This is the only reason I don’t use Kavita currently. Heck, even if it were added under Kavita+ I’d pay for that feature by itself.

**Comment by DieselTech** [comment]
For those following along at home, you can reference this issue to see what is needed to get ODIC started: https://github.com/Kareadita/Kavita/issues/2518

**Comment by darkf3n1x** [comment]
This would be a fantastic feature to have! This way we could utilize plex auth or self hosted authentik.

**Comment by Aloe-recite** [comment]
any update on this? really missing a SSO option right now!

**Reply by DieselTech** [reply]
You can read through #2518 to get the overall idea of where this is at. Based on the stats we pulled from active installs reporting in, the average Kavita install has 1.2 users per instance. That isn't enough to warrant the massive amount of time it would take to make this feature. That isn't to say we won't ever support it, but just that the kavita team won't be the ones to create it. 

tl;dr - If the community wants OIDC then the community is going to be the one to create the implementation of it. If it meets the minimum standard we set forth then it could get added to the codebase.

**Reply by micchickenburger** [reply]
Sounds like selection bias to me.  How do you know installs reporting in are representative of your population?  That's the first thing I turn off when I install new software.

**Reply by DieselTech** [reply]
It's fine to turn off the basic stat reporting if you want, but we do use those numbers to figure out what to work on when it comes to features. As stated in our Wiki, you can see the stats that are being reported here https://github.com/Kareadita/KavitaStats || https://wiki.kavitareader.com/troubleshooting/faq/#q-does-kavita-collect-any-data-on-me

So if you turn off that info, that effectively removes your voice from the pool to be heard. It isn't like we are tracking what your filenames are or what you are reading. 

The stats have shown us that the average number of users a kavita server has is 1.2 per install. That just isn't enough of the userbase to warrant the dev time this would take compared to adding other features and improving the software in ways that will affect more people. 

As I said above, it isn't like we are outright saying "No" here. It just won't be the Kavita team implementing this feature. People keep getting mad this feature doesn't exist, yet they aren't contributing any code to make it happen either.

**Reply by bnerickson** [reply]
It's certainly well within the contributors' rights to focus on other features especially when introducing SSO support is going to be a real headache.  However, I think focusing on the number of users per-install ignores the other benefits that SSO provides.  Namely, that the account I use to login can have its password changed in the identity management system and be reflected in other applications that tie into it.

If every application author said, "we're unwilling to work on supporting SSO because the number of users per install is less-than 2," then I'd have to update my password in (presently) 8 different self-hosted applications.  Another potentially flawed argument: perhaps the number of installed users is 1.2 because administrators are unwilling to create and manage multiple accounts in Kavita if they can't do so in their identity management system instead :) ?

Regardless, I have no problem with the regular contributors focusing on other features and hope maintainers understand most people aren't angry that there's no SSO.  Hopefully someone takes a crack at it at some point.

Edit: Actually, now that I think about it I only have 1 user in my install (admin) precisely because I don't want to manage a separate user account outside of an identity management system.

**Reply by micchickenburger** [reply]
@DieselTech I think this is where the selection bias comes into play: I think it's the case that folks who are privacy conscious or inherently distrustful are more likely to operate an IdM provider.  If you find that the volume surrounding this issue is incongruous with the statistic, perhaps the statistic is not giving you the information you're looking for.

Further, as @bnerickson so aptly states, IdM systems have value even for just one user.  Today I couldn't log into Kavita because of the dozen or so apps I'm currently self-hosting, Kavita is the only one that doesn't support SSO and I can't remember what I set the credentials.

I'd be happy to contribute, but given the need for a significant refactor, one or more people with existing experience with the codebase would likely be better choices.

I do have a suggestion that would address this issue for at least some of us:  Proxy authentication.  When enabled, get the username and email address from the HTTP headers `X-Authenticated-User` and `X-Authenticated-User-Email`, as examples.  A reverse proxy can authenticate a user before allowing access to the upstream service, ie Kavita.  Then, Kavita can match the header values with an existing account and treat it as authenticated.  For security, an IP allow list is essential to avoid manipulation of HTTP requests, and documentation must also call out that care must be taken to ensure Kavita is ONLY accessible through the reverse proxy.

**Reply by majora2007** [reply]
Let me chime in (again) with my stance. 99% of Kavita's development comes from me, spare the few PRs or help from team members. 

As I stated, while I understand OIDC is a major ask from the community, I personally am not interested in working on it. I believe it will be difficult integrating with the existing Authentication system. Think I'm wrong? Do the investigation, open a PR and let's work together to bring it within Kavita. 

If you're not participating in the stats, that's fine. If you think that features to be included are solely backed by stats, then you haven't been involved in the community Kavita has. I drive a lot of community ideas into Kavita (like the comic library overhaul requiring refactoring over 30K lines of core code). But at the end of the day, I have a final say. I want to work on what I think will holistically make Kavita better. I don't use OIDC, I use a password manager that autofills and Kavita has refresh tokens, so I literally never have the need for OIDC. 

So to recap, you can discuss all you want, but it's falling on deaf ears. It's a one person project and development is at my discretion. **I'm open, so if you want it, do it.**

**Comment by ZLoth** [comment]
Add this administrator who would love to integrate Authentik with Kavita to make user management easier.

**Comment by majora2007** [comment]
We just released a special, on-off build of our OIDC Implementation for Kavita. This build is only useable via Docker (`docker pull jvmilazz0/kavita:canary`). 

Please help us by testing and leaving feedback either as a response to **this message** or in discord on [this](https://discord.com/channels/821879810934439936/1395760358173577348/1395760412393214112) discord thread.

**Reply by stancubed** [reply]
Thank you for this! Quick note: the URL that the OpenID Connect page points to currently results in a 404: https://wiki.kavitareader.com/guides/admin-settings/open-id-connect/
I think I've got my Pocket-ID set up properly and Kavita seems to be calling out to it, but I seem to be getting an OIDC missing-email error. Email is being provided from Pocket-ID though.

```
[Kavita] [2025-07-27 15:57:02.661 -05:00  43] [Error] Microsoft.AspNetCore.Authentication.JwtBearer.JwtBearerHandler Exception occurred while processing message.

Kavita.Common.KavitaException: errors.oidc.missing-email

   at API.Services.OidcService.LoginOrCreate(HttpRequest request, ClaimsPrincipal principal) in /home/runner/work/Kavita/Kavita/API/Services/OidcService.cs:line 72

   at API.Extensions.IdentityServiceExtensions.OidcClaimsPrincipalConverter(TokenValidatedContext ctx) in /home/runner/work/Kavita/Kavita/API/Extensions/IdentityServiceExtensions.cs:line 159

   at Microsoft.AspNetCore.Authentication.JwtBearer.JwtBearerHandler.HandleAuthenticateAsync()

[Kavita] [2025-07-27 15:57:02.662 -05:00  43] [Error] API.Middleware.ExceptionMiddleware There was an exception

Kavita.Common.KavitaException: errors.oidc.missing-email

   at API.Services.OidcService.LoginOrCreate(HttpRequest request, ClaimsPrincipal principal) in /home/runner/work/Kavita/Kavita/API/Services/OidcService.cs:line 72

   at API.Extensions.IdentityServiceExtensions.OidcClaimsPrincipalConverter(TokenValidatedContext ctx) in /home/runner/work/Kavita/Kavita/API/Extensions/IdentityServiceExtensions.cs:line 159

   at Microsoft.AspNetCore.Authentication.JwtBearer.JwtBearerHandler.HandleAuthenticateAsync()

   at Microsoft.AspNetCore.Authentication.JwtBearer.JwtBearerHandler.HandleAuthenticateAsync()

   at Microsoft.AspNetCore.Authentication.AuthenticationHandler`1.AuthenticateAsync()

   at Microsoft.AspNetCore.Authentication.AuthenticationService.AuthenticateAsync(HttpContext context, String scheme)

   at Microsoft.AspNetCore.Authentication.AuthenticationHandler`1.AuthenticateAsync()

   at Microsoft.AspNetCore.Authentication.AuthenticationService.AuthenticateAsync(HttpContext context, String scheme)

   at Microsoft.AspNetCore.Authentication.AuthenticationMiddleware.Invoke(HttpContext context)

   at Microsoft.AspNetCore.ResponseCaching.ResponseCachingMiddleware.Invoke(HttpContext httpContext)

   at Microsoft.AspNetCore.ResponseCompression.ResponseCompressionMiddleware.InvokeCore(HttpContext context)

   at API.Middleware.SecurityEventMiddleware.InvokeAsync(HttpContext context) in /home/runner/work/Kavita/Kavita/API/Middleware/SecurityMiddleware.cs:line 26

   at API.Middleware.ExceptionMiddleware.InvokeAsync(HttpContext context) in /home/runner/work/Kavita/Kavita/API/Middleware/ExceptionMiddleware.cs:line 18
```

**Reply by majora2007** [reply]
Just a reminder that this is in canary. The wiki will not link correctly until the code is merged into nightly.

**Reply by stancubed** [reply]
Good reminder! Might be worth calling that out since it's included in the canary build and/or providing an alternate link in the canary build so that testers can work from the information you have.

**Reply by 2o1o0** [reply]
Hello,

I just switched to this canary build, I've setup my client in keycloak as I usually do, but when I try to connect I first get a 502 when its trying to redirect me to `https://mykavitadomain/signin-oidc` and then if I refresh I can see the error `Message contains error: 'invalid_grant', error_description: 'Code not valid', error_uri: 'error_uri is null'. ` on a kavita login screen error toaster.
Any idea what I could have missed?

In the logs :
 ```
 [Kavita] [2025-07-29 00:18:17.243 +02:00 49] [Error] Microsoft.AspNetCore.Authentication.OpenIdConnect.OpenIdConnectHandler Exception occurred while processing message.
Microsoft.IdentityModel.Protocols.OpenIdConnect.OpenIdConnectProtocolException: Message contains error: 'invalid_grant', error_description: 'Code not valid', error_uri: 'error_uri is null'.
at Microsoft.AspNetCore.Authentication.OpenIdConnect.OpenIdConnectHandler.RedeemAuthorizationCodeAsync(OpenIdConnectMessage tokenEndpointRequest)
 ```
<details>
<img width="284" height="302" alt="image" src="https://github.com/user-attachments/assets/eaa030ea-26dc-44a2-996e-74bf9875bc6a" />
<img width="462" height="872" alt="image" src="https://github.com/user-attachments/assets/a1ea4ae9-4a67-43eb-8484-d06ac34899b4" />
</details>

**Reply by majora2007** [reply]
Please join the discord. We have active testing there, we aren't actively supporting here. 

https://discord.com/channels/821879810934439936/1395760358173577348

**Comment by majora2007** [comment]
This feature has been delivered in the v0.8.7.7 nightly and will be a part of v0.8.8 stable release.
